xdsl/parser.py

```diff
@@ -1399,7 +1291,8 @@ def try_parse_ref_attr(self) -> SymbolRefAttr | None:
         else:
             return None
 
-    def try_parse_builtin_int_attr(self) -> IntegerAttr | None:
+    def try_parse_builtin_int_attr(
+            self) -> IntegerAttr[IntegerType | IndexType] | None:
```

Comment:
Is there a way to differentiate among the two to get stronger type guarantees? Maybe through two more functions, one that returns an IntegerAttr of IntegerType, the other of IndexType. And then, when using those in custom parsing, you should get better type guarantees...

Maybe this is out of scope for this PR though: Curious to hear your thoughts :)